sql/colexec: TestSpillingQueue failed

[(sql/colexec).TestSpillingQueue failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1965214&tab=buildLog) on [master@809829bfe7ff27a610fa78f409fae658b9f2d9d9](https://github.com/cockroachdb/cockroach/commits/809829bfe7ff27a610fa78f409fae658b9f2d9d9):

Fatal error:
```
panic: runtime error: integer divide by zero [recovered]
	panic: runtime error: integer divide by zero

```

Stack:
```
goroutine 9561 [running]:
testing.tRunner.func1(0xc000150600)
	/usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0x2b9dac0, 0x4b06630)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/sql/colexec.newSpillingQueue(0xc00053e420, 0xc0001c8028, 0x1, 0x1, 0xf4169, 0x14ee078f18d0, 0xc0002a4000, 0x2ea3a02, 0x7, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue.go:87 +0x341
github.com/cockroachdb/cockroach/pkg/sql/colexec.newRewindableSpillingQueue(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue.go:120
github.com/cockroachdb/cockroach/pkg/sql/colexec.TestSpillingQueue.func1(0xc000150600)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue_test.go:79 +0x3e1
testing.tRunner(0xc000150600, 0xc0120aaa50)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestSpillingQueue/Rewindable/MemoryLimit=1.1_MiB/DiskQueueCacheMode=0/AlwaysCompress=true/NumBatches=372

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestSpillingQueue PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSpillingQueue.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colexec).TestSpillingQueue failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1966763&tab=buildLog) on [master@533095b4f6d74439d6c11c0b57f5c7b80937fb5c](https://github.com/cockroachdb/cockroach/commits/533095b4f6d74439d6c11c0b57f5c7b80937fb5c):

Fatal error:
```
panic: runtime error: integer divide by zero [recovered]
	panic: runtime error: integer divide by zero

```

Stack:
```
goroutine 9276 [running]:
testing.tRunner.func1(0xc00018fe00)
	/usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0x2b9dac0, 0x4b06630)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/cockroachdb/cockroach/pkg/sql/colexec.newSpillingQueue(0xc0004d2660, 0xc0005a8020, 0x1, 0x1, 0xffffffffffff2800, 0x147b2e88c558, 0xc00056eb40, 0x2ea3a02, 0x7, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue.go:87 +0x341
github.com/cockroachdb/cockroach/pkg/sql/colexec.newRewindableSpillingQueue(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue.go:120
github.com/cockroachdb/cockroach/pkg/sql/colexec.TestSpillingQueue.func1(0xc00018fe00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue_test.go:79 +0x3e1
testing.tRunner(0xc00018fe00, 0xc008f0aa50)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   TestSpillingQueue/Rewindable/MemoryLimit=10_KiB/DiskQueueCacheMode=1/AlwaysCompress=false/NumBatches=228

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestSpillingQueue PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSpillingQueue.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colmem/allocator.go
+++ b/pkg/sql/colmem/allocator.go
@@ -283,6 +283,7 @@ const (
 	SizeOfTime = int(unsafe.Sizeof(time.Time{}))
 	// SizeOfDuration is the size of a single duration.Duration value.
 	SizeOfDuration = int(unsafe.Sizeof(duration.Duration{}))
+	sizeOfDatum    = int(unsafe.Sizeof(tree.Datum(nil)))
 )
 
 // SizeOfBatchSizeSelVector is the size (in bytes) of a selection vector of
@@ -336,8 +337,12 @@ func EstimateBatchSizeBytes(vecTypes []*types.T, batchLength int) int {
 		case types.IntervalFamily:
 			acc += SizeOfDuration
 		case typeconv.DatumVecCanonicalTypeFamily:
-			size, _ := tree.DatumTypeSize(t)
-			acc += int(size)
+			// In datum vec we need to account for memory underlying the struct
+			// that is the implementation of tree.Datum interface (for example,
+			// tree.DBoolFalse) as well as for the overhead of storing that
+			// implementation in the slice of tree.Datums.
+			implementationSize, _ := tree.DatumTypeSize(t)
+			acc += int(implementationSize) + sizeOfDatum
 		default:
 			colexecerror.InternalError(fmt.Sprintf("unhandled type %s", t))
 		}

